### ISSUE TITLE
DateField/TimeField are broken when full-width chatacter is input.

### ISSUE THREAD

**Issue by uraway** [adobe/react-spectrum#3739]
# 🐛 Bug Report 

![image](https://user-images.githubusercontent.com/15242484/201310502-52a48e85-d12d-4076-8328-b63c372b0349.png)
https://codesandbox.io/s/recursing-meadow-20eibw?file=/src/App.js

Input a fullwidth character like `あ` for each field after `new Time(11, 11)` is set for TimeField.

## 🤔 Expected Behavior

DateField/TimeField should ignore full-width input like NumberField

## 😯 Current Behavior

When I input full-width characters, the values displayed on the field seem to stop working.

## 💁 Possible Solution

Is there a workaround to exclude non-numeric values on the DateField/TimeField?

## 🔦 Context

I tried to use `onKeyDown/onKeyUp` events to prevent users from full-width characters input, but these events are not working?

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | 3.22.0
| Browser          |
| Operating System |

**Comment by LFDanLu** [comment]
Reproduced locally, will need to investigate why the NumberField behavior is different from our date/time components

**Comment by LFDanLu** [comment]
from what I can tell NumberField is different because it is a textfield and but the DateField segments are `contentEditable` spans. I tried tinkering around to see if I could get rid of the issue locally, and it seems to be https://github.com/adobe/react-spectrum/blob/1aa29e8237fcc1b25e58b63d7ba51488e69464cf/packages/%40react-aria/datepicker/src/useDateSegment.ts#L306-L310 and https://github.com/adobe/react-spectrum/blob/1aa29e8237fcc1b25e58b63d7ba51488e69464cf/packages/%40react-aria/datepicker/src/useDateSegment.ts#L326 which cause the subsequent crashes when using ArrowUp/Down after attempting to input a full width character. Ideally we'd be able to prevent those from being set I imagine when typing any invalid character but doing so is proving tricky

**Comment by LFDanLu** [comment]
Did some more digging with @devongovett, the crash is mainly isolated to RSP DateField at the moment due to  https://github.com/adobe/react-spectrum/blob/a5876d28643bfe2bb42c7bc12cf0de34532f3426/packages/%40react-spectrum/datepicker/src/DatePickerSegment.tsx#L66 where we remove the placeholder upon typing, causing the textContent restoration to happen on the parent ref rather than the original placeholder. We can attempt to find all the text areas that will be modified before input and reset them post input, but there is still a problem with the IME dropdown not showing up as you type due to this resetting it seems.

**Comment by snowystinger** [comment]
@LFDanLu did you fix this in your Number Parser PR?

**Comment by LFDanLu** [comment]
nope, we looked at that issue in tandem with this one but couldn't fix the above issue of the crashes/placeholders restorations
